integrations/gitlab2/main.py

```diff
@@ -0,0 +1,207 @@
+    webhook_url = ""
+    semaphore = asyncio.Semaphore(10)
+
+
+    async for projects in gitlab_client.get_projects():
+        logger.info(f"Received project batch with {len(projects)}")
+        tasks = [
+            create_webhook_with_semaphore(
+                semaphore,
+                gitlab_client,
+                project['id'],
+                webhook_url
+            )
+            for project in projects
+        ]
+
+        results = await asyncio.gather(*tasks, return_exceptions=True)
+
+        for result in results:
+            if isinstance(result, Exception):
+                logger.error(f"Error occurred during webhook creation: {result}")
+
+        yield projects
+
+@ocean.on_resync(ObjectKind.MERGE_REQUEST)
+async def on_merge_request_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE :
+    gitlab_client = init_client()
+    async for merge_requests in gitlab_client.get_merge_requests():
+        yield merge_requests
+
+@ocean.on_resync(ObjectKind.ISSUE)
+async def on_issue_resync(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE :
+    gitlab_client = init_client()
+    async for issues in gitlab_client.get_issues():
+        logger.info(f"Received issues batch with {len(issues)}")
+        yield issues
+
+
+@ocean.router.post("/webhook")
+async def handle_webhook(request: Request):
+    event_type = request.headers.get('X-Gitlab-Event')
+    payload = await request.json()
+    await process_webhook_event(event_type, payload)
+
+
+async def create_webhook_with_semaphore(semaphore: asyncio.Semaphore, gitlab_client: GitlabClient, project_id: int, webhook_url: str ):
+    async with semaphore:
+        await gitlab_client.create_project_webhook(project_id)
+
+async def process_webhook_event(event_type: str, payload: dict):
+    gitlab_client = init_client()
+
+    if event_type == "Issue Hook":
+        await handle_issue_event(payload, gitlab_client)
+    elif event_type == "Merge Request Hook":
+        await handle_merge_request_event(payload, gitlab_client)
+    elif event_type == "Push Hook":
+        await handle_push_event(payload, gitlab_client)
+    elif event_type == "Project Hook":
+        await handle_project_event(payload, gitlab_client)
+    elif event_type == "Group Hook":
+        await handle_group_event(payload, gitlab_client)
+    else:
+        logger.warning(f"Unhandled event type: {event_type}")
+
+async def handle_issue_event(payload: dict, gitlab_client: GitlabClient):
+    issue_attributes = payload.get('object_attributes', {})
+    issue_id = issue_attributes.get('id')
+    project_id = payload.get('project', {}).get('id')
+    action = issue_attributes.get('action')  # 'open', 'close', 'reopen', 'update'
+    logger.info(f"Issue {action}: {issue_id}")
+
+    if action in ['open', 'reopen', 'update']:
+        # Fetch the latest issue data
+        issue = await gitlab_client.get_single_issue(project_id, issue_attributes.get('iid'))
+        await ocean.register_raw(ObjectKind.ISSUE, [issue])
```

Comment:
these are great but I will prefer we keep the main.py clean and move these implementation to a separate file